opentelemetry-instrumentation/src/opentelemetry/instrumentation/_labeler/_internal/__init__.py

```diff
@@ -0,0 +1,168 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextvars
+import threading
+from typing import Any, Dict, Optional, Union
+
+# Context variable to store the current labeler
+_labeler_context: contextvars.ContextVar[Optional["Labeler"]] = (
+    contextvars.ContextVar("otel_labeler", default=None)
+)
+
+
+class Labeler:
+    """
+    Stores custom attributes for the current request in context.
+
+    This feature is experimental and unstable.
+    """
+
+    def __init__(self):
+        self._lock = threading.Lock()
+        self._attributes: Dict[str, Union[str, int, float, bool]] = {}
+
+    def add(self, key: str, value: Union[str, int, float, bool]) -> None:
+        """
+        Add a single attribute to the labeler.
+
+        Args:
+            key: The attribute key
+            value: The attribute value (must be a primitive type)
+        """
+        with self._lock:
+            self._attributes[key] = value
+
+    def add_attributes(
+        self, attributes: Dict[str, Union[str, int, float, bool]]
+    ) -> None:
+        """
+        Add multiple attributes to the labeler.
+
+        Args:
+            attributes: Dictionary of attributes to add
+        """
+        with self._lock:
+            self._attributes.update(attributes)
+
+    def get_attributes(self) -> Dict[str, Union[str, int, float, bool]]:
+        """
+        Returns a copy of all attributes added to the labeler.
+        """
+        with self._lock:
+            return self._attributes.copy()
+
+    def clear(self) -> None:
+        with self._lock:
+            self._attributes.clear()
+
+    def __len__(self) -> int:
+        with self._lock:
+            return len(self._attributes)
+
+
+def get_labeler() -> Labeler:
+    """
+    Get the Labeler instance for the current request context.
+
+    If no Labeler exists in the current context, a new one is created
+    and stored in the context.
+
+    Returns:
+        Labeler instance for the current request, or a new empty Labeler
+        if not in a request context
+    """
+    labeler = _labeler_context.get()
+    if labeler is None:
+        labeler = Labeler()
+        _labeler_context.set(labeler)
+    return labeler
+
+
+def set_labeler(labeler: Labeler) -> None:
+    """
+    Set the Labeler instance for the current request context.
+
+    Args:
+        labeler: The Labeler instance to set
+    """
+    _labeler_context.set(labeler)
+
+
+def clear_labeler() -> None:
+    """
+    Clear the Labeler instance from the current request context.
+    """
+    _labeler_context.set(None)
+
+
+def get_labeler_attributes() -> Dict[str, Union[str, int, float, bool]]:
+    """
+    Get attributes from the current labeler, if any.
+
+    Returns:
+        Dictionary of custom attributes, or empty dict if no labeler exists
+    """
+    labeler = _labeler_context.get()
+    if labeler is None:
+        return {}
```

Comment:
```suggestion
        return MappingProxyType({})
```
maybe use MappingProxyType here and in get_attributes return MappingProxyType(self._attributes)

Reply:
Updated in [c86a538](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3689/commits/c86a538ce80cf0832f31ff404adc4b60ba739cc1)